src/nxt_http_static.c

```diff
@@ -571,7 +629,9 @@ nxt_http_static_send_ready(nxt_task_t *task, void *obj, void *data)
         /* Not a file. */
         nxt_file_close(task, f);
 
-        if (nxt_slow_path(!nxt_is_dir(&fi))) {
+        if (nxt_slow_path(!nxt_is_dir(&fi)
+                          || shr->start[shr->length - 1] == '/'))
```

Comment:
If `shr` ends with `/`, it can return directly with 404 before opening file.

Reply:
The rationale here is:

If `shr` ends with `/`, we use (append) `index`; otherwise we don't.  That's okay, and not a failure condition.

Then, independently of having used `index` or not, we open the file.

If it's a regular file, we proceed normally.
If it's not a regular file, but not a directory, we had this 404.
If it's a directory, we were redirecting (301) to `${r->path}/`.

Since the index was hardcoded to `"index.html"`, it was hard to get that to be a file (unless the server did something stupid such as `mdkir index.html` or maybe symlinked the index file to a dir), so that was enough.

Now that we have a configurable index, it's easier to have it be a directory: `"index": ""` or `"index": {}` will result in nothing appended to `shr`, and therefore, if `shr` ends with `/`, `r->path` will also end with `/`.

Since we are appending nothing, the problem is really deeper, because the redirection, if followed would just append `/` to the URI, which would result in a loop that ends up in something like `${r->path}//////`.

The condition to fail, independently of when we apply it, should be:

**if**
we have used `index` (the easiest thing to test this is "if `shr` ends with `/`)
**and**
 the file is a directory (this can be tested in two ways; see below).

The 2 ways we can test the second part of the condition are:
- by opening the file and checking that it's a directory.  This is 100% sure.
- by checking that either `index` is empty or it ends with `/`.  This is good enough, but it's not perfect, because it would not account for the case that `"index": "dir"`, which would then return 301 and a Location of `$shr/dir/` (which would be inconsistent with the rest).

The first way is the one I decided to implement because of:

It consistently returns 404 for directories after appending `index`.
Since having a directory _after_ appending index means that something is wrong in the server (because why would that happen in a sane server?), I don't want to optimize for that case at the cost of having some larger code that will make the common (and correct) case a bit slower.

Reply:
@tippexs @lcrilly @hongzhidao Ignoring performance, are there any security implications of choosing one of the above or the other?

Reply:
> It consistently returns 404 for directories after appending index.

I agree.

> If shr ends with /, it can directly enter nxt_http_static_next() before opening file.

Sorry that I didn't make it clear. Take a look again.

```
diff -r bb4e9ca76ef9 src/nxt_http_static.c
--- a/src/nxt_http_static.c     Fri May 13 19:46:08 2022 +0800
+++ b/src/nxt_http_static.c     Fri May 13 20:49:44 2022 +0800
@@ -347,6 +348,9 @@ nxt_http_static_send_ready(nxt_task_t *t
         fname = ctx->share.start;
     }

+    // if fname ends with "/\0",
+    // nxt_http_static_next(task, r, ctx, NXT_HTTP_NOT_FOUND);
+
     nxt_memzero(&file, sizeof(nxt_file_t));

     file.name = fname;
```

Reply:
Yes, I also considered that.  It's a mix of both ways.

The thing with this is that this optimizes for the cases that `"index": {}` or `"index": ""` or `"index": "dir/"`.
At the cost of adding 2 lines of code.

I consider all of those 3 cases to be broken.  Especially, the last two seem completely useless to me.  The first one may be useful to an automated configuration generator that may be adding or removing lines, and may end up with an empty array; but still I don't think we should optimize for that corner case, if it adds 2 LOC to the common case, do you?